src/vos/vos_dtx.c

```diff
@@ -935,14 +1306,155 @@ vos_dtx_check(daos_handle_t coh, struct dtx_id *dti)
 	return do_vos_dtx_check(coh, dti, NULL);
 }
 
-void
+int
 vos_dtx_commit_internal(struct vos_container *cont, struct dtx_id *dtis,
-			int count)
+			int count, daos_epoch_t epoch)
 {
-	int	i;
+	struct vos_cont_df		*cont_df = cont->vc_cont_df;
+	struct umem_instance		*umm = vos_cont2umm(cont);
+	struct vos_dtx_scm_blob		*dsb;
+	struct vos_dtx_scm_blob		*dsb_prev;
+	umem_off_t			 dsb_off;
+	struct vos_dtx_cmt_ent_df	*dce_df;
+	int				 slots = 0;
+	int				 cur = 0;
+	int				 rc = 0;
+	int				 rc1 = 0;
+	int				 i;
+	int				 j;
+
+	dsb = umem_off2ptr(umm, cont_df->cd_dtx_committed_tail);
+	if (dsb != NULL)
+		slots = dsb->dsb_cap - dsb->dsb_count;
+
+	if (slots == 0)
+		goto new_blob;
+
+	umem_tx_add_ptr(umm, &dsb->dsb_count, sizeof(dsb->dsb_count));
+
+again:
+	if (slots > count)
+		slots = count;
+
+	count -= slots;
+
+	if (slots > 1) {
+		D_ALLOC(dce_df, sizeof(*dce_df) * slots);
+		if (dce_df == NULL)
+			return -DER_NOMEM;
+	} else {
+		dce_df = &dsb->dsb_commmitted_data[dsb->dsb_count];
+	}
+
+	for (i = 0, j = 0; i < slots; i++, cur++) {
+		struct vos_dtx_cmt_ent	*dce = NULL;
+
+		rc = vos_dtx_commit_one(cont, &dtis[cur], epoch, &dce);
+		if (rc1 == 0)
+			rc1 = rc;
+
+		if (dce != NULL) {
+			if (slots == 1)
```

Comment:
you could have a dce_df in stack to avoid allocation for single entry, so we can avoid "if" statement just because dce is in different memory types.

Reply:
We only allocate temporary @dce_df for (slots > 1) case.